CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs

```diff
@@ -882,6 +917,16 @@ public double As(UnitSystem unitSystem)
             return ToUnit(unit, DefaultConversionFunctions);
         }}
 
+        /// <summary>
+        ///     Converts this {_quantity.Name} to another {_quantity.Name} with the unit representation <paramref name=""unit"" /> and returns its see <cref name=""QuantityValue"" />.
+        /// </summary>
+        /// <param name=""unit"">The unit to convert to.</param>
+        /// <returns>A {_quantity.Name} with the specified unit.</returns>
+        public QuantityValue ToQuantity({_unitEnumName} unit)
```

Comment:
The reason I'm trying to avoid option D is that it's error prone looking forward. 

Assume you had written (to be on the "safe" side):
```csharp
// Option D - No method, let consumer cast instead.
return MassFlow.FromKilogramsPerSecond((double)bsfc.KilogramsPerJoule * (double)power.Watts);
```
and now you decide that `BrakeSpecificFuelConsumption` (the left hand argument) should use `decimal` as base type as well. There won't be any errors or warnings, but the result of the above calculation will be inexact and nobody would notice. It gets even uglier when we'd once have more base types. 

With the options A-C, the multiplication actually uses `QuantityValue.operator*()` to perform the operation. The left hand side is implicitly converted to QuantityValue (it wouldn't hurt to do it explicitly) and we have one place to define the behavior when multiplying values of different types with each other. Otherwise, each of these multiplications or divisions would need to specify what type the result of e.g. decimal*double should be. 

About A-C: I don't have a big preference there, but we should be aware that this scenario also affects user code if he performs an operation on units that's not predefined in the library. Hence A is probably the easiest to use.

Reply:
>  There won't be any errors or warnings, but the result of the above calculation will be inexact and nobody would notice. 

I agree with this sentiment, that we don't want calculations to change without anyone noticing. 

However;

1. I think it is more likely to run into changes to calculations without noticing when `QuantityValue` casts values internally. It is not intuitive for the consumer to know if they are getting double or decimal precision on their calculations.
2. You argue `we have one place to define the behavior when multiplying values of different types with each other`. This is convenient for us, but I still think it is a trap for the consumer.
3. Having to call `ToQuantityValue()` is not very intuitive. Getting a compile error on trying to multiple `double` with `decimal` is highly discoverable, although yes it requires an explicit cast.
4. It is with good reason that .NET also fails to compile in this case, instead of trying to automatically cast one way or the other.

Reply:
That sounds reasonable. 

What about we go with solution B (from the above list), which performs an explicit cast from double/decimal to QuantityValue, so the option to use `QuantityValue.operator*` is possible if desired. 

Actually, we do not need to do anything to allow explicit casts. These are always possible. Internally, we should avoid them though, to prevent the error described above.

Reply:
Looking back at this with fresh eyes, I still lean towards solutions C or D, basically forcing the consumer to make an informed choice when mixing arithmetic of decimal vs double.

@lipchev @tmilnthorp what do you think?

```c#
// Option A - minor rename, discoverable, but must be added to all 100+ quanities
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.ToQuantityValue(PowerUnit.Watt));

// Option B - New `QuantityValue` property, less discoverable and the extra value property seems a bit weird until you know why it is useful.
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.ToUnit(PowerUnit.Watt).QuantityValue);

// Option C - Extension method on `decimal` to convert to `QuantityValue`, less discoverable, but 1 vs 100 methods on binary size
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.Watts.ToQuantityValue());

// Option D - No method, let consumer cast instead.
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule*(double)power.Watts);
```

Reply:
I think option D is most consistent with the change of the `IQuantity.Value` type: anybody migrating to v5 would have to get used to casting on the deconstruction side: `var value = (double) quantity.Value` (or something like `quantity.Value.ToDouble()`), so I guess the same logic should apply on the construction side.

The particular line that you've highlighted however points to the other issue- around the question of what types are selected when multiplying two different types, with the options being:

- Option A, B, C: resulting in a Double * Decimal => Decimal => Double (if we follow the operator in the **public** `QuantityValue` class)
- Option D: resulting in Double * Double => Double  // note that the behavior is slightly different from what we'd get by doing `MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.Watts)` by hand (without casting)

On the opposite side of the operation:
A, B, C => ~ `return SpecificEnergy.FromJoulesPerKilogram(power.Watts.ToQuantityValue() / massFlow.KilogramsPerSecond);` // (Decimal / Double) => Decimal => Double
vs
D =>   `return SpecificEnergy.FromJoulesPerKilogram((double)power.Watts / massFlow.KilogramsPerSecond);` // (Double/ Double) => Double : again slightly different from power.Watts / massFlow.KilogramsPerSecond (result here is Decimal)

In short, option D is the `operation strategy™` where operands are cast to the **expected** result type prior to performing the operations, while the other ones let the resulting type vary - with a bias for Decimals (with possible overflows). 

My point is that unless we want to support **both** double & decimal as a possible `_value` for a given quantity then there is little point in relying on operators defined in the `QuantityValue` (especially since we'd be overriding all but the trivial cases).

Reply:
> My point is that unless we want to support both double & decimal as a possible _value for a given quantity then there is little point in relying on operators defined in the QuantityValue (especially since we'd be overriding all but the trivial cases).

That's a fair point. I expected that multiplying a `double` with a `decimal` by first converting the double to decimal would always return the correct result. But that's not true, that multiplication can always loose accuracy or significant digits. So it's probably better to always leave the decision to the user as to how exactly to perform the conversion.